src/internal/storage/fileset/transforms.go

```diff
@@ -3,27 +3,46 @@ package fileset
 import (
 	"context"
 	"io"
+	"strings"
 
 	"github.com/pachyderm/pachyderm/v2/src/internal/storage/fileset/index"
 )
 
 var _ FileSet = &indexFilter{}
 
 type indexFilter struct {
-	pred func(idx *index.Index) bool
-	x    FileSet
+	fs        FileSet
+	predicate func(*index.Index) bool
+	full      bool
 }
 
-// NewIndexFilter filters x using pred
-func NewIndexFilter(x FileSet, pred func(idx *index.Index) bool) FileSet {
-	return &indexFilter{x: x, pred: pred}
+// NewIndexFilter filters fs using predicate.
+func NewIndexFilter(fs FileSet, predicate func(idx *index.Index) bool, full ...bool) FileSet {
+	idxf := &indexFilter{
+		fs:        fs,
+		predicate: predicate,
+	}
+	if len(full) > 0 {
+		idxf.full = full[0]
+	}
+	return idxf
 }
 
-func (fil *indexFilter) Iterate(ctx context.Context, cb func(File) error, _ ...bool) error {
-	return fil.x.Iterate(ctx, func(fr File) error {
-		idx := fr.Index()
-		if fil.pred(idx) {
-			return cb(fr)
+func (idxf *indexFilter) Iterate(ctx context.Context, cb func(File) error, _ ...bool) error {
```

Comment:
Does the last argument need to be propagated through?

Reply:
Yeah, it does. I'll fix this for all of the transforms. Ideally we would be able to get rid of this method of specifying the deletive set some time in the near future.

Reply:
This is evidence of the optional argument introducing a bug in practice, so I'll suggest removing it again.

Reply:
Yeah, I don't think the deletive set should go through the same function as a normal iteration. I don't think we should do that in this PR though.